proposals/2020-07-28_keyring-cmm/proposal.md

```diff
@@ -0,0 +1,198 @@
+Second, it serves as a safe default that can be used for most use cases.
+
+While there is currently no conflict in these purposes,
+this might not always be true.
+In the future we might want to update the behavior customers get when they choose some "default"
+CMM option.
+
+One way to address this would be to directly add the new desired behavior
+to the current Default CMM implementation,
+however this makes our provided CMMs less composable.
+Keeping the CMM composable is important because it allows users to assert desired properties
+[by construction](../../tenets.md#correct-by-construction).
+Because it is important to retain this base implementation,
+we should define a CMM whose sole purpose is to be this base implementation.
+We will call this CMM implementation the Keyring CMM.
+For implementations which need to support master key provider, we will also define a
+Master Key Provider CMM.
+
+Similar to how we define a set of algorithm suites and define a default algorithm suite,
+we should similarly have a set of provided CMM implementations
+and a Default CMM which represents a particular configuration of one of those CMMs
+(possibly composed with other CMMs).
+We will define that the Default CMM is "the Keyring CMM".
+Just as an example, if we wanted to provide caching as a default behavior,
+the Default CMM could be defined as "the Caching CMM configured with a Keyring CMM."
+By defining the default in this way, we have greater flexibility to update the default
+CMM that customers use in the future without losing the composability of these implementations.
+
+Such a default MUST NOT allow any options other than the keyring to be configured in order
+to allow us to more easily maintain backwards compatability with possible Default CMM updates.
+
+## Drawbacks
+
+This change SHOULD NOT have any drawbacks.
+
+## Security Implications
+
+This change SHOULD NOT have any security implications.
+
+## Operational Implications
+
+This change SHOULD NOT have any operational implications.
+
+This change MUST maintain behavior for customers who are using a Default CMM.
+
+## Guide-level Explanation
+
+This change defines the [Keyring CMM](#keyring-cmm) and
+the [Master Key Provider CMM](#master-key-provider-cmm)
+as base CMM implementations that can be used "as is" or composed with other CMM.
+
+This change also defines a [Default CMM](#default-cmm) as a specific configuration of a CMM
+that the AWS Encryption SDK provides as a safe default.
+
+Customers SHOULD either use the [Default CMM](#default-cmm)
+or compose a CMM using the AWS Encryption SDK's provided CMM implementations
+(including the Keyring CMM).
+
+### Default CMM
+
+The Default CMM is a specific configuration of a
+CMM implementation provided by the AWS Encryption SDK.
+
+The specific CMM configuration describes a safe default that serves most use cases.
+
+The CMM configuration defined by the Default CMM is the [Keyring CMM](#keyring-cmm) as is
```

Comment:
I'm not sure I understand. If you are suggesting that we be able to initialize the KeyringCMM with a "default alg" option that overrides the default defined in the alg suite, then that would be changing the current behavior of the Default CMM and is out of scope.

Reply:
I was suggesting that the Keyring CMM takes an option for "default alg",
and then when the Keyring CMM is created in the Default CMM,
then the option would be what we define as a default.

So the ESDK default algorithm is defined in the Default CMM and not in the Keyring CMM.

Reply:
What if the keyring CMM just requires that the request identify the algorithm suite?

I'm loathe to squeeze opinions about algorithm suite into a CMM that SHOULD only care about wrangling keyrings.

Reply:
I understand what you're proposing now. This would mean that the Default CMM is no longer *just* a configuration. Unless we also want to move to the "default to the default alg CMM" logic out of the Keyring CMM and into a new CMM.

If we were to do something like this, I agree with Matt's approach, the KeyringCMM should just fail if it doesn't get an alg suite. I'm not sure if splitting up the logic this way is useful or not though.

Reply:
This makes sense, but the keyring CMM should be easy to use on its own.
If customers chose to use _only_ the keyring CMM and inject that into the ESDK, it would not work in the default case.
Maybe that is a good thing though.

What I would like is that the default algorithm suite for the ESDK is defined in the Default CMM, not the keyring CMM.

Reply:
> What I would like is that the default algorithm suite for the ESDK is defined in the Default CMM, not the keyring CMM.

Agreed. If you're opting out of the default then you don't get any of the defaults. I think this is reasonable.

Reply:
For clarity, we do not want to have the Keyring CMM take a REQUIRED option "default alg"?

Reply:
I might be misunderstanding this conversation, but near the end of the doc, we say "This CMM MUST NOT offer any additional features beyond the composed CMM created above." To me, this means the default CMM MAY offer the same configuration options as the thing it is configuring. However, it MAY additionally support sane defaults.

Reply:
After offline discussion:

- It is important that the Default CMM MUST represent an expression of default values without accompanying logic
  - This defines all defaults for encryption and decryption in one place, and additionally makes these defaults more extendable for any future code that may use CMMs but is not necessarily our current Encrypt/Decrypt client APIs. 
  - By ensuring this is a strict configuration, we ensure that each behavior is encapsulated in a composable CMM which can be reused by customers, making it easy to compose CMMs which opt out of certain default behaviors or reuse default behaviors within another configuration.
  - This also makes updating the default behaviors easy, and able to be done in a way to preserves composability of our behaviors.
- As such, the default algorithm suite is defined by the Default CMM
  - Customers MUST be able to get this default algorithm suite statically
- One more behavior of the KeyringCMM in this version of the spec is that it defaults to the default algorithm. This is not desired and this behavior should be encapsulated outside the Keyring CMM implementation.
  - The Keyring CMM MUST require an algorithm suite
- We should add a new CMM implementation DefaultRequestValues (name TBD) that is solely responsible for setting default values in the CMM's materials requests, and then passes that request to an underlying CMM.
  - For this proposal, this should ONLY be setting some default algorithm suite for the get encryption materials request.
- The Default CMM should thus represent the composition of this new DefaultRequestValues CMM and a Keyring CMM.
- We need to add examples for composing CMMs together.

I will update the doc with this motivation and behavior.